site/en/docs/extensions/reference/runtime/index.md

```diff
@@ -2,4 +2,79 @@
 api: runtime
 ---
 
-<!-- Intentionally blank -->
+## Examples
+
+### Use [`getURL`][getURL] to add an extension image to a page {: #example-getURL }
+
+Extensions can access their own resources using relative paths, but in order for the extension's
+assets to appear in a web page, you will need to specify the full URL of the asset (and expose it
+using the [web_accessible_resources][war] field).
+
+This example shows how a [content script][content] could add an image in the extension's package to
+the page that the content script has been [injected][content-inject] into.
+
+```js
+//// content.js
+
+{ // Block used to avoid setting global variables
+  let img = document.createElement('img');
+  img.src = chrome.runtime.getURL('logo.png');
+  document.body.append(img);
+}
+```
+
+### Use runtime messaging to communicate between background and content scripts {: #example-background-content-messaging }
+
+A basic example of content script injected into a page and background script passing messages to
+each other. This exchange is modeled on a [TCP handshake][handshake].
+
+```js
+//// content.js ////
+
+// 1. Send the background a message requesting receipt confirmation.
+console.log('[content] sending content-syn');
+chrome.runtime.sendMessage('content-syn');
+
+chrome.runtime.onMessage.addListener((message, sender) => {
+  if (message === 'bg-syn-ack') {
+    // 3. BG syn-ack received, acknowledged receipt.
+    console.log('[content] received bg-syn-ack');
+    console.log('[content] sending content-ack');
+    chrome.runtime.sendMessage('content-ack');
```

Comment:
This is an anti-pattern. This approach doesn't utilize the one-time messaging mechanism properly: extensions should use sendResponse() provided to the callback as the third parameter as shown in the [message passing](https://developer.chrome.com/docs/extensions/mv3/messaging/#simple) documentation.

Your examples might be suitable for port-based messaging but even then they'll be incorrect in case there are two cycles going on in two tabs. You need to keep a map of tabId-to-messageId - such map is maintained by the built-in one-time messaging automatically.

Reply:
Forked this discussion into issue #647 as I don't think this is a blocker for merging this PR.

Reply:
Well, this code is typical for people who started learning the extensions API, and that's definitely an antipattern, and it will introduce bugs in certain cases so I don't see any reason why you'd want to add it...

Reply:
After some additional reflection I think I now understand your concerns and agree that we shouldn't merge this PR with the current example. I was too focusing on emulating an arbitrary request flow rather than real world scenarios where different extensions contexts need to pass messages. I'll try to get an update in for this tonight.

Reply:
Yay!

BTW, for your backlog, there's another problematic example - the first one in the [message passing](https://developer.chrome.com/docs/extensions/mv3/messaging/#simple) article I've linked. It uses chrome.tabs.query() to find the active tab. The problem is that people copypaste it in the wrong places such as tabs.onUpdated or webNavigation listeners instead of using the event's tab or tabId parameter so their code works only if the activity happens in the currently focused tab. I guess this situation can be rectified by adding an example for a listener with a short explanation.

Reply:
This example has been replaced with a simpler content script message and background script `sendResponse()` call.